core/src/main/scala/org/apache/spark/scheduler/Task.scala

```diff
@@ -106,7 +106,22 @@ private[spark] abstract class Task[T](val stageId: Int, var partitionId: Int) ex
     if (interruptThread && taskThread != null) {
       taskThread.interrupt()
     }
-  }  
+  }
+
+  override def hashCode(): Int = {
+    val state = Seq(stageId, partitionId)
+    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
+  }
+
+  def canEqual(other: Any): Boolean = other.isInstanceOf[Task[T]]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: Task[_] =>
+      (that canEqual this) &&
```

Comment:
`that` is already a `Task` here; is the `canEqual` method adding anything? the check has no dependency on the generic type `T`.